### PR TITLE
Introduce BufferedLogForwarder

### PR DESCRIPTION
**Pull request by zwass**
`BufferedLogForwarder` is a base class for external log forwarding schemes
that require buffering and retries. It generalizes the logic from the
existing `TLSLogForwarderRunner`.

The existing `TLSLoggerPlugin` and `TLSLogForwarderRunner` are refactored to
demonstrate the use of this new base class.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by marpaia** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2036/merge (Job: 2594) failed :-1:

**Comment by zwass** [comment]
womp. Looking.

**Comment by zwass** [comment]
Ah, so I was running the tests with `./build/darwin/osquery/osquery_additional_tests --gtest_filter="LoggerPluginUtilTests.*"` which has them running sequentially and everything is all good. I have a feeling the test runner is parallelizing them and that's breaking things because of a shared DB. I thought that calling `initTesting()` was handling that in `SetUp()` but I guess not?

File: osquery/logger/plugins/logger_plugin_util.h [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#pragma once
```
Comment (marpaia) [inline osquery/logger/plugins/logger_plugin_util.h]: I would vote to move this to `osquery/logger/buffer.cpp` or something. I like having only plugins in `plugins` and the name is pretty long as it is now.

Comment (zwass) [inline osquery/logger/plugins/logger_plugin_util.h]: Yeah I'd love to shorten the name. Maybe `BufferedLogger`?

Comment (marpaia) [inline osquery/logger/plugins/logger_plugin_util.h]: I don't mind the class name being `BufferedLogForwarder`, I was more referring to the filename itself. We don't have perfect 1:1 parity in class name to file name (nor do we have much consistency around keeping one class in a file, etc).

File: osquery/logger/plugins/logger_plugin_util.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  }
+}
+
+Status BufferedLogForwarder::logString(const std::string& s) {
+  std::string index = genLogIndex(true);
+  return setDatabaseValue(kLogs, index, s);
+}
+
+Status BufferedLogForwarder::logStatus(const std::vector<StatusLogLine>& log) {
+  // Append decorations to status (unique to TLS logger).
```
Comment (theopolis) [inline osquery/logger/plugins/logger_plugin_util.cpp]: I guess this note no longer applies. :P

File: osquery/logger/plugins/logger_plugin_util.h [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  std::string genLogIndex(bool results);
+
+  /// Seconds between flushing logs
+  std::chrono::seconds log_period_;
+
+  /// Max number of logs to flush per check
+  size_t max_log_lines_;
+
+  /// Hold an incrementing index for buffering logs
+  size_t log_index_;
```
Comment (theopolis) [inline osquery/logger/plugins/logger_plugin_util.h]: You can initialize this here:
```cpp
size_t log_index_{0};
```

Then you wont have to repeat the initialization in each constructor.

File: osquery/logger/plugins/tests/logger_plugin_util_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+#include "osquery/core/test_util.h"
+#include "osquery/logger/plugins/logger_plugin_util.h"
+
+namespace osquery {
+
+using namespace testing;
+
+class LoggerPluginUtilTests : public Test {
+ public:
+  void SetUp() {
```
Comment (theopolis) [inline osquery/logger/plugins/tests/logger_plugin_util_tests.cpp]: Ditch this override. `initTesting()` should be called once, in the tester's `main()`.

File: osquery/logger/plugins/tls.h [hunk 5]
```diff
@@ -82,19 +67,10 @@ class TLSLoggerPlugin : public LoggerPlugin {
...
-   * @brief Hold an auto-incrementing offset for buffered logs.
-   *
-   * Logs are buffered to a backing store until they can be flushed to a TLS
-   * endpoint (based on latency/retry/etc options). Buffering uses a UNIX time
-   * second precision for indexing and ordering. log_index_ helps prevent
-   * collisions by appending an auto-increment counter.
-   */
-  size_t log_index_{0};
+  // Forwarder that buffers/sends logs. Runs in a Dispatcher thread.
+  std::shared_ptr<TLSLogForwarder> forwarder_ = nullptr;
```
Comment (theopolis) [inline osquery/logger/plugins/tls.h]: Not sure what the pros/cons are, but we've typically stuck to initializer list syntax for member variable initialization:
```cpp
std::shared_ptr<TLSLogForwarder> forwarder_{nullptr};
```

Also, if you use `///` the comment is assumed to be the `@breif` string for doxygen.

**Comment by osqueryer** [comment]
Build origin/pr/2036/merge (Job: 2597) failed :-1:

**Comment by osqueryer** [comment]
Build origin/pr/2036/merge (Job: 2598) failed :-1:

**Comment by osqueryer** [comment]
Build origin/pr/2036/merge (Job: 2599) failed :-1:

**Comment by zwass** [comment]
It looks like the one test failure we're getting is due to an error on the build host:
```
FATAL: Could not checkout af22d003f01dd8f353d13559a0719845f8a9cd02
hudson.plugins.git.GitException: Could not checkout af22d003f01dd8f353d13559a0719845f8a9cd02
...
```
I'm going to re-push to set off the build again and hopefully it was just a transient error.

**Comment by zwass** [comment]
Are there any integration tests for the TLS logger? I feel much better about the unit testing situation with all of the mocked testing of the buffering/retry logic in the `BufferedLogForwarder`, but running an integration test would make me feel real nice.

**Comment by marpaia** [comment]
@zwass `osquery/logger/plugins/tests/tls_logger_tests.cpp` is pretty integration-y (perhaps unfortunately). `TLSServerRunner::start();` runs `tools/tests/test_http_server.py` which is an actual TLS server and it uses that to test a bunch of stuff.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2036/merge (Job: 2606) failed :-1:

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build origin/pr/2036/merge (Job: 2607) is good! :+1:

**Pull request merged**